From c8e993dd71c2e7d830dc860a1fe5720c2433cc21 Mon Sep 17 00:00:00 2001
From: Alexander Duyck <alexander.h.duyck@redhat.com>
Date: Fri, 31 Oct 2014 19:06:15 -0400
Subject: [net] ipv4: fix a potential use after free in ip_tunnel_core.c

Message-id: <20141031190615.8251.1190.stgit@ahduyck-vm-rhel7>
Patchwork-id: 98653
O-Subject: [RHEL7.1 net PATCH 1/7] ipv4: fix a potential use after free in ip_tunnel_core.c
Bugzilla: 1151886 1152368
RH-Acked-by: Florian Westphal <fwestpha@redhat.com>
RH-Acked-by: Hannes Frederic Sowa <hannes@redhat.com>
RH-Acked-by: Jiri Pirko <jpirko@redhat.com>

Bugzilla: 1151886
Brew Build: 8179440

Backport of upstream commit: 1245dfc8cadb258386fcd27df38215a0eccb1f17

 commit 1245dfc8cadb258386fcd27df38215a0eccb1f17
 Author: Li RongQing <roy.qing.li@gmail.com>
 Date:   Fri Oct 17 16:53:23 2014 +0800

    ipv4: fix a potential use after free in ip_tunnel_core.c

    pskb_may_pull() maybe change skb->data and make eth pointer oboslete,
    so set eth after pskb_may_pull()

    Fixes:3d7b46cd("ip_tunnel: push generic protocol handling to ip_tunnel module")
    Cc: Pravin B Shelar <pshelar@nicira.com>
    Signed-off-by: Li RongQing <roy.qing.li@gmail.com>
    Acked-by: Pravin B Shelar <pshelar@nicira.com>
    Signed-off-by: David S. Miller <davem@davemloft.net>

Signed-off-by: Alexander Duyck <alexander.h.duyck@redhat.com>
Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/ipv4/ip_tunnel_core.c b/net/ipv4/ip_tunnel_core.c
index 2e7692c..6daee8c 100644
--- a/net/ipv4/ip_tunnel_core.c
+++ b/net/ipv4/ip_tunnel_core.c
@@ -93,11 +93,12 @@ int iptunnel_pull_header(struct sk_buff *skb, int hdr_len, __be16 inner_proto)
  skb_pull_rcsum(skb, hdr_len);
 
  if (inner_proto == htons(ETH_P_TEB)) {
-  struct ethhdr *eh = (struct ethhdr *)skb->data;
+  struct ethhdr *eh;
 
   if (unlikely(!pskb_may_pull(skb, ETH_HLEN)))
    return -ENOMEM;
 
+  eh = (struct ethhdr *)skb->data;
   if (likely(ntohs(eh->h_proto) >= ETH_P_802_3_MIN))
    skb->protocol = eh->h_proto;
   else
-- 
1.7.1